unstructured/ingest/connector/fsspec.py

```diff
@@ -209,16 +220,104 @@ def _list_files(self):
                 if v.get("size") > 0
             ]
 
+    def _decompress_file(self, filename: str) -> str:
+        head, tail = os.path.split(filename)
+        if any(filename.endswith(ext) for ext in zip_file_extensions):
+            for ext in zip_file_extensions:
+                if tail.endswith(ext):
+                    tail = tail[: -(len(ext))]
+                    break
+            path = os.path.join(head, f"{tail}-zip-uncompressed")
+            logger.info(f"extracting zip {filename} -> {path}")
+            with zipfile.ZipFile(filename) as zfile:
+                zfile.extractall(path=path)
+            return path
+        elif any(filename.endswith(ext) for ext in tar_file_extensions):
+            for ext in tar_file_extensions:
+                if tail.endswith(ext):
+                    tail = tail[: -(len(ext))]
+                    break
+
+            path = os.path.join(head, f"{tail}-tar-uncompressed")
+            logger.info(f"extracting tar {filename} -> {path}")
+            with tarfile.open(filename, "r:gz") as tfile:
+                tfile.extractall(path=path)
+            return path
+        else:
+            raise ValueError(
+                "filename {} not a recognized compressed extension: {}".format(
+                    filename,
+                    ", ".join(zip_file_extensions + tar_file_extensions),
+                ),
+            )
+
+    def _process_compressed_doc(self, doc: FsspecIngestDoc) -> t.List[BaseIngestDoc]:
+        # Download the raw file to local
+        doc.get_file()
+        path = self._decompress_file(filename=str(doc.filename))
+        new_read_configs = copy.copy(self.read_config)
+        new_partition_configs = copy.copy(self.partition_config)
+        relative_path = path.replace(self.read_config.download_dir, "")
+        if relative_path[0] == "/":
+            relative_path = relative_path[1:]
+        new_partition_configs.output_dir = os.path.join(
+            self.partition_config.output_dir,
+            relative_path,
+        )
+
+        local_connector = LocalSourceConnector(
+            connector_config=SimpleLocalConfig(
+                input_path=path,
+                recursive=True,
+            ),
+            read_config=new_read_configs,
+            partition_config=new_partition_configs,
+        )
+        local_connector.initialize()
+        return local_connector.get_ingest_docs()
```

Comment:
does this need to be FsspecIngestDoc specific? could this be a mixin so other connectors can reuse?